From 54a7c43b1d18ca6ca0ae91f306e30697cb52b7f0 Mon Sep 17 00:00:00 2001
From: Emergi Idan <idan.emergi@intel.com>
Date: Fri, 1 May 2020 00:21:42 +0300
Subject: [PATCH] PONRTSYS-7747: fix qos fw logger

Use infinite read and write indexes to handle
wraparound properly

Signed-off-by: Emergi Idan <idan.emergi@intel.com>
---
 drivers/net/ppv4/pp-api/pp-misc/pp_misc_debugfs.c  | 11 +++++----
 drivers/net/ppv4/pp-api/pp-qos-mgr/pp_qos_fw.c     | 26 +++++++++++++++++-----
 .../net/ppv4/pp-api/pp-qos-mgr/pp_qos_uc_defs.h    |  4 ++--
 3 files changed, 29 insertions(+), 12 deletions(-)

diff --git a/drivers/net/ppv4/pp-api/pp-misc/pp_misc_debugfs.c b/drivers/net/ppv4/pp-api/pp-misc/pp_misc_debugfs.c
index 6dee17837f20..8e352d80d6c5 100644
--- a/drivers/net/ppv4/pp-api/pp-misc/pp_misc_debugfs.c
+++ b/drivers/net/ppv4/pp-api/pp-misc/pp_misc_debugfs.c
@@ -35,19 +35,22 @@ static phys_addr_t ppv4_reg_base;
 
 void regs_dump_rd(struct seq_file *f)
 {
-	seq_puts(f, "<physical address> <num regs (optional)>\n");
+	seq_puts(f,
+		 "<physical address> <num regs (optional)> <regs per line>\n");
 }
 
 void regs_dump_wr(char *cmd_buf, void *data)
 {
 	unsigned long addr;
 	u32 num_regs = 1;
+	u32 regs_per_line = 4;
 	u32 reg, i, n = 0;
 	char kbuf[128];
 	void *virt;
 	s32 ret;
 
-	if (sscanf(cmd_buf, "%lx %u", &addr, &num_regs) != 2) {
+	ret = sscanf(cmd_buf, "%lx %u %u", &addr, &num_regs, &regs_per_line);
+	if (ret < 2) {
 		ret = kstrtoul(cmd_buf, 0, &addr);
 		if (ret) {
 			pr_err("kstrtou32 err (%d)\n", ret);
@@ -60,7 +63,7 @@ void regs_dump_wr(char *cmd_buf, void *data)
 
 	kbuf[0] = '\0';
 	for (i = 0; i < num_regs; i++, addr += 4) {
-		if (!(i % 4)) {
+		if (!(i % regs_per_line)) {
 			if (i)
 				pr_info("%s\n", kbuf);
 			n = scnprintf(kbuf, sizeof(kbuf), "0x%08lx:\t", addr);
@@ -585,7 +588,7 @@ static struct debugfs_file drv_files[] = {
 static struct debugfs_file hal_files[] = {
 	{"read",    &PP_DEBUGFS_FOPS(regs_rd)},
 	{"write32", &PP_DEBUGFS_FOPS(regs_wr)},
-#if IS_ENABLED(CONFIG_PPV4_LGM)	
+#if IS_ENABLED(CONFIG_PPV4_LGM)
 	{"global",          &PP_DEBUGFS_FOPS(hal_global)},
 	{"pps",             &PP_DEBUGFS_FOPS(hal_pps)},
 	{"resources",       &PP_DEBUGFS_FOPS(resources)},
diff --git a/drivers/net/ppv4/pp-api/pp-qos-mgr/pp_qos_fw.c b/drivers/net/ppv4/pp-api/pp-qos-mgr/pp_qos_fw.c
index 727445be0166..3483c9bb75d1 100644
--- a/drivers/net/ppv4/pp-api/pp-qos-mgr/pp_qos_fw.c
+++ b/drivers/net/ppv4/pp-api/pp-qos-mgr/pp_qos_fw.c
@@ -3163,6 +3163,7 @@ void print_fw_log(struct pp_qos_dev *qdev)
 {
 	char msg[PPV4_QOS_LOGGER_MSG_SIZE];
 	u32  num_messages;
+	u32  max_messages;
 	u32  msg_idx;
 	u32  msg_read_count;
 	u32  *msg_word;
@@ -3172,18 +3173,31 @@ void print_fw_log(struct pp_qos_dev *qdev)
 
 	pp_cache_invalidate(qdev->hwconf.fw_logger_ddr_virt,
 			    qdev->hwconf.fw_logger_sz);
+	max_messages =
+		(qdev->hwconf.fw_logger_sz - PPV4_QOS_LOGGER_HEADROOM_SIZE) /
+		PPV4_QOS_LOGGER_MSG_SIZE;
 	addr = (unsigned long)(qdev->hwconf.fw_logger_ddr_virt);
 	num_messages = *(u32 *)(addr + PPV4_QOS_LOGGER_WRITE_IDX_OFFSET);
 	msg_read_count = *(u32 *)(addr + PPV4_QOS_LOGGER_READ_IDX_OFFSET);
 
+	/* check if wraparound occur */
+	if ((num_messages - msg_read_count) > max_messages)
+		msg_read_count = num_messages - max_messages;
+
 	cur = (unsigned long)(qdev->hwconf.fw_logger_ddr_virt) +
 		PPV4_QOS_LOGGER_RING_OFFSET;
 
-	pr_info("addr is %#lx num of messages is %u, read count is %u\n",
-		cur, num_messages, msg_read_count);
-
-	for (msg_idx = msg_read_count; msg_idx < num_messages; ++msg_idx) {
-		for (msg_word_idx = 0 ; msg_word_idx < 32 ; ++msg_word_idx) {
+	pr_info("FW Total messages Printed : %u\n", num_messages);
+	pr_info("Host total messages read  : %u\n", msg_read_count);
+	pr_info("Messages to print         : %u\n",
+		num_messages - msg_read_count);
+	pr_info("Log buffer size           : %u\n", max_messages);
+	pr_info("Buffer base               : %#lx\n", cur);
+	pr_info("==============================================\n");
+
+	for (; msg_read_count < num_messages; ++msg_read_count) {
+		msg_idx = msg_read_count % max_messages;
+		for (msg_word_idx = 0; msg_word_idx < 32; ++msg_word_idx) {
 			msg_word = (u32 *)(msg) + msg_word_idx;
 			msg_addr = cur +
 				(PPV4_QOS_LOGGER_MSG_SIZE * msg_idx) +
@@ -3197,7 +3211,7 @@ void print_fw_log(struct pp_qos_dev *qdev)
 	}
 
 	/* When done write the correct read count */
-	*(u32 *)(addr + PPV4_QOS_LOGGER_READ_IDX_OFFSET) = num_messages;
+	*(u32 *)(addr + PPV4_QOS_LOGGER_READ_IDX_OFFSET) = msg_read_count;
 	pp_cache_writeback((void *)addr + PPV4_QOS_LOGGER_READ_IDX_OFFSET, 4);
 }
 
diff --git a/drivers/net/ppv4/pp-api/pp-qos-mgr/pp_qos_uc_defs.h b/drivers/net/ppv4/pp-api/pp-qos-mgr/pp_qos_uc_defs.h
index 6fbcb5856e62..392002758ce6 100644
--- a/drivers/net/ppv4/pp-api/pp-qos-mgr/pp_qos_uc_defs.h
+++ b/drivers/net/ppv4/pp-api/pp-qos-mgr/pp_qos_uc_defs.h
@@ -11,8 +11,8 @@
 
 /* UC version */
 #define UC_VERSION_MAJOR (1)
-#define UC_VERSION_MINOR (12)
-#define UC_VERSION_BUILD (2)
+#define UC_VERSION_MINOR (13)
+#define UC_VERSION_BUILD (0)
 
 #define QOS_MAX_PREDECESSORS        (6)
 #define QOS_AQM_CONTEXT_MAX_QUEUES  (8)
